plugins/telemetry-otel/src/main/java/org/opensearch/telemetry/tracing/sampler/ProbabilisticSampler.java

```diff
@@ -37,23 +47,56 @@ public ProbabilisticSampler(TelemetrySettings telemetrySettings) {
         this.telemetrySettings = Objects.requireNonNull(telemetrySettings);
         this.samplingRatio = telemetrySettings.getSamplingProbability();
         this.defaultSampler = Sampler.traceIdRatioBased(samplingRatio);
+        this.actionBasedSampler = new HashMap<>();
```

Comment:
@devagarwal1803 three questions there: 
 - Why we bundle action sampler into `ProbabilisticSampler`? I think it should separate `TransportActionSampler` that deals only with actions, not probabilities (this is why we have `ProbabilisticSampler`)
 - I think the samplers should be composable: fe  `ProbabilisticSampler` -> `TransportActionSampler` or  `TransportActionSampler` -> `ProbabilisticSampler`, depending on the intent
 - I think it the samplers could also be used independently: fe only  `ProbabilisticSampler` (like right now) or `TransportActionSampler` 

At the larger scope: how the transport actions are validated? we cannot accept any "garbage" here?

Reply:
+1 on defining the samplers independent and should be chained through composite sampler.

Reply:
Making samplers independent & chained through composite/common sampler makes sense. But I have a small question on the approach. Since we are defining the probability of sampling based on action also as well as general sampling probability also, so let's support a scenario: 

*Setup* - "I have a sampling defined as 0.5 for an action `dummy_action` & 0.5 for general probabilistic sampling also". Suppose there are 4 request for `dummy_action`.

*Working* - As per the action based sampling 2 of them will be sampled, now the remaining 2 which are not sampled will be passed to probabilistic sampler, out of which it will also sample 1 request.

*Expected Result* - 2 request out of 4 should be sampled.

*Actual Result* - 3 request are sampled out of 4.

Reply:
> Setup - "I have a sampling defined as 0.5 for an action dummy_action & 0.5 for general probabilistic sampling also". Suppose there are 4 request for dummy_action.

I think we should not use probably per action at all (at least at this stage), the probabilistic sampler could be composed with, but the probability would be one for all

Reply:
So, from my understanding our Composite sampler will call the TransportAction. Then it will check if there is an action present or not in the request & sample based on it (with same probability for all actions). 
If no, then it will return `RECORD_ONLY` (object of Sampling result) & will call Probablistic sampler.

Please do let me know if my understand is correct.

Reply:
> Please do let me know if my understand is correct.

The logic depends how samplers are being composed, you've described one flow, a few more examples (`parent` is always there, `RequestSampler`):
1. -> parent -> Probability -> Action -> ...
2. -> parent -> Action -> Probability -> ...
3. -> parent -> Probability
4. -> parent -> Action